From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Danila Tikhonov <danila@jiaxyga.com>
Date: Fri, 1 Mar 2024 19:05:54 +0300
Subject: [PATCH 042/166] WIP! power: supply: Add rude driver for QGauge

This is battery fuel gauge driver for pmic5 (PMI632/PM6150/PM7250B)
and pmic-lite (PM2250).

Signed-off-by: Danila Tikhonov <danila@jiaxyga.com>
---
 drivers/power/supply/Kconfig   |  10 ++
 drivers/power/supply/Makefile  |   1 +
 drivers/power/supply/qcom_qg.c | 258 +++++++++++++++++++++++++++++++++
 3 files changed, 269 insertions(+)
 create mode 100644 drivers/power/supply/qcom_qg.c

diff --git a/drivers/power/supply/Kconfig b/drivers/power/supply/Kconfig
index b89ef40df7f8..93082381ae9f 100644
--- a/drivers/power/supply/Kconfig
+++ b/drivers/power/supply/Kconfig
@@ -217,6 +217,16 @@ config BATTERY_QCOM_BATTMGR
 	  which is used on modern Qualcomm platforms to provide battery and
 	  power supply information.
 
+config BATTERY_QCOM_QG
+	tristate "Qualcomm PMIC QGauge driver"
+	depends on MFD_SPMI_PMIC
+	depends on IIO
+	help
+	  Say Y here to enable the Qualcomm Technologies, Inc. QGauge driver
+	  which uses the periodic sampling of the battery voltage and current
+	  to determine the battery state-of-charge (SOC) and supports other
+	  battery management features.
+
 config BATTERY_OLPC
 	tristate "One Laptop Per Child battery"
 	depends on OLPC_EC
diff --git a/drivers/power/supply/Makefile b/drivers/power/supply/Makefile
index f3c0d1006c1a..baf5bd2a6a7f 100644
--- a/drivers/power/supply/Makefile
+++ b/drivers/power/supply/Makefile
@@ -38,6 +38,7 @@ obj-$(CONFIG_BATTERY_LEGO_EV3)	+= lego_ev3_battery.o
 obj-$(CONFIG_BATTERY_LENOVO_YOGA_C630) += lenovo_yoga_c630_battery.o
 obj-$(CONFIG_BATTERY_PMU)	+= pmu_battery.o
 obj-$(CONFIG_BATTERY_QCOM_BATTMGR)	+= qcom_battmgr.o
+obj-$(CONFIG_BATTERY_QCOM_QG)	+= qcom_qg.o
 obj-$(CONFIG_BATTERY_OLPC)	+= olpc_battery.o
 obj-$(CONFIG_BATTERY_SAMSUNG_SDI)	+= samsung-sdi-battery.o
 obj-$(CONFIG_BATTERY_S2MU005)	+= s2mu005-battery.o
diff --git a/drivers/power/supply/qcom_qg.c b/drivers/power/supply/qcom_qg.c
new file mode 100644
index 000000000000..8bd15e1f842d
--- /dev/null
+++ b/drivers/power/supply/qcom_qg.c
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (c) 2024, Danila Tikhonov <danila@jiaxyga.com>
+ */
+
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/of_address.h>
+#include <linux/platform_device.h>
+#include <linux/of.h>
+#include <linux/power_supply.h>
+#include <linux/iio/consumer.h>
+#include <linux/regmap.h>
+
+/* SRAM */
+#define QG_SRAM_BASE	0xb600
+
+/* BATT offsets */
+#define QG_S2_NORMAL_AVG_V_DATA0_REG	0x80 /* 2-byte 0x80-0x81 */
+#define QG_S2_NORMAL_AVG_I_DATA0_REG	0x82 /* 2-byte 0x82-0x83 */
+#define QG_LAST_ADC_V_DATA0_REG		0xc0 /* 2-byte 0xc0-0xc1 */
+#define QG_LAST_ADC_I_DATA0_REG		0xc2 /* 2-byte 0xc2-0xc3 */
+
+/* SRAM offsets */
+#define QG_SDAM_OCV_OFFSET		0x4c /* 4-byte 0x4c-0x4f */
+#define QG_SDAM_LEARNED_CAPACITY_OFFSET	0x68 /* 2-byte 0x68-0x69 */
+
+struct qcom_qg_chip {
+	struct device *dev;
+	struct regmap *regmap;
+	unsigned int base;
+
+	struct iio_channel *batt_therm_chan;
+
+	struct power_supply *batt_psy;
+	struct power_supply_battery_info *batt_info;
+};
+
+static int qcom_qg_get_current(struct qcom_qg_chip *chip, u8 offset, int *val)
+{
+	s16 temp;
+	u8 readval[2];
+	int ret;
+
+	ret = regmap_bulk_read(chip->regmap, chip->base + offset, readval, 2);
+	if (ret) {
+		dev_err(chip->dev, "Failed to read current: %d\n", ret);
+		return ret;
+	}
+
+	temp = (s16)(readval[1] << 8 | readval[0]);
+	*val = div_s64((s64)temp * 152588, 1000);
+
+	return 0;
+}
+
+static int qcom_qg_get_voltage(struct qcom_qg_chip *chip, u8 offset, int *val)
+{
+	int ret, temp;
+	u8 readval[2];
+
+	ret = regmap_bulk_read(chip->regmap, chip->base + offset, readval, 2);
+	if (ret) {
+		dev_err(chip->dev, "Failed to read voltage: %d\n", ret);
+		return ret;
+	}
+
+	temp = readval[1] << 8 | readval[0];
+	*val = div_u64((u64)temp * 194637, 1000);
+
+	return 0;
+}
+
+/*
+ * Yes, this function simply calculates the capacity based on
+ * the current voltage. This will be rewritten in the future.
+ */
+static int qcom_qg_get_capacity(struct qcom_qg_chip *chip, int *val)
+{
+	int ret, voltage_now;
+	int voltage_min = chip->batt_info->voltage_min_design_uv;
+	int voltage_max = chip->batt_info->voltage_max_design_uv;
+
+	ret = qcom_qg_get_voltage(chip,
+				QG_S2_NORMAL_AVG_V_DATA0_REG, &voltage_now);
+	if (ret) {
+		dev_err(chip->dev, "Failed to get current voltage: %d\n", ret);
+		return ret;
+	}
+
+	if (voltage_now <= voltage_min)
+		*val = 0;
+	else if (voltage_now >= voltage_max)
+		*val = 100;
+	else
+		*val = (((voltage_now - voltage_min) * 100) /
+						(voltage_max - voltage_min));
+
+	return 0;
+}
+
+static enum power_supply_property qcom_qg_props[] = {
+	POWER_SUPPLY_PROP_STATUS,
+	POWER_SUPPLY_PROP_TECHNOLOGY,
+	POWER_SUPPLY_PROP_VOLTAGE_MAX_DESIGN,
+	POWER_SUPPLY_PROP_VOLTAGE_MIN_DESIGN,
+	POWER_SUPPLY_PROP_VOLTAGE_NOW,
+	POWER_SUPPLY_PROP_VOLTAGE_AVG,
+	POWER_SUPPLY_PROP_VOLTAGE_OCV,
+	POWER_SUPPLY_PROP_CURRENT_NOW,
+	POWER_SUPPLY_PROP_CURRENT_AVG,
+	POWER_SUPPLY_PROP_CHARGE_FULL_DESIGN,
+	POWER_SUPPLY_PROP_CHARGE_FULL,
+	POWER_SUPPLY_PROP_CAPACITY,
+	POWER_SUPPLY_PROP_TEMP,
+};
+
+static int qcom_qg_get_property(struct power_supply *psy,
+				enum power_supply_property psp,
+				union power_supply_propval *val)
+{
+	struct qcom_qg_chip *chip = power_supply_get_drvdata(psy);
+	int ret;
+
+	switch (psp) {
+	case POWER_SUPPLY_PROP_STATUS:
+		val->intval = POWER_SUPPLY_STATUS_UNKNOWN;
+		break;
+	case POWER_SUPPLY_PROP_TECHNOLOGY:
+		val->intval = POWER_SUPPLY_TECHNOLOGY_LIPO;
+		break;
+	case POWER_SUPPLY_PROP_VOLTAGE_MAX_DESIGN:
+		val->intval = chip->batt_info->voltage_max_design_uv;
+		break;
+	case POWER_SUPPLY_PROP_VOLTAGE_MIN_DESIGN:
+		val->intval = chip->batt_info->voltage_min_design_uv;
+		break;
+	case POWER_SUPPLY_PROP_VOLTAGE_NOW:
+		ret = qcom_qg_get_voltage(chip,
+				QG_LAST_ADC_V_DATA0_REG, &val->intval);
+		break;
+	case POWER_SUPPLY_PROP_VOLTAGE_AVG:
+		ret = qcom_qg_get_voltage(chip,
+				QG_S2_NORMAL_AVG_V_DATA0_REG, &val->intval);
+		break;
+	case POWER_SUPPLY_PROP_VOLTAGE_OCV:
+		ret = regmap_raw_read(chip->regmap,
+			QG_SRAM_BASE + QG_SDAM_OCV_OFFSET, &val->intval, 4);
+		break;
+	case POWER_SUPPLY_PROP_CURRENT_NOW:
+		ret = qcom_qg_get_current(chip,
+				QG_LAST_ADC_I_DATA0_REG, &val->intval);
+		break;
+	case POWER_SUPPLY_PROP_CURRENT_AVG:
+		ret = qcom_qg_get_current(chip,
+				QG_S2_NORMAL_AVG_I_DATA0_REG, &val->intval);
+		break;
+	case POWER_SUPPLY_PROP_CHARGE_FULL_DESIGN:
+		val->intval = chip->batt_info->charge_full_design_uah;
+		break;
+	case POWER_SUPPLY_PROP_CHARGE_FULL:
+		ret = regmap_raw_read(chip->regmap, QG_SRAM_BASE +
+				QG_SDAM_LEARNED_CAPACITY_OFFSET, &val->intval, 2);
+		if (!ret) val->intval *= 1000; /* mah to uah */
+		break;
+	case POWER_SUPPLY_PROP_CAPACITY:
+		ret = qcom_qg_get_capacity(chip, &val->intval);
+		break;
+	case POWER_SUPPLY_PROP_TEMP:
+		ret = iio_read_channel_processed
+					(chip->batt_therm_chan, &val->intval);
+		break;
+	default:
+		dev_err(chip->dev, "invalid property: %d\n", psp);
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static struct power_supply_desc batt_psy_desc = {
+	.name = "qcom_qg",
+	.type = POWER_SUPPLY_TYPE_BATTERY,
+	.properties = qcom_qg_props,
+	.num_properties = ARRAY_SIZE(qcom_qg_props),
+	.get_property = qcom_qg_get_property,
+};
+
+static int qcom_qg_probe(struct platform_device *pdev)
+{
+	struct qcom_qg_chip *chip;
+	struct power_supply_config psy_cfg = {};
+	int ret;
+
+	chip = devm_kzalloc(&pdev->dev, sizeof(*chip), GFP_KERNEL);
+	if (!chip)
+		return -ENOMEM;
+
+	chip->dev = &pdev->dev;
+
+	/* Regmap */
+	chip->regmap = dev_get_regmap(pdev->dev.parent, NULL);
+	if (!chip->regmap)
+		return dev_err_probe(chip->dev, -ENODEV,
+				     "Failed to locate the regmap\n");
+
+	/* Get base address */
+	ret = device_property_read_u32(chip->dev, "reg", &chip->base);
+	if (ret < 0)
+		return dev_err_probe(chip->dev, ret,
+				     "Couldn't read base address\n");
+
+	/* ADC for thermal channel */
+	chip->batt_therm_chan = devm_iio_channel_get(&pdev->dev, "batt-therm");
+	if (IS_ERR(chip->batt_therm_chan))
+		return dev_err_probe(chip->dev, PTR_ERR(chip->batt_therm_chan),
+				     "Couldn't get batt-therm IIO channel\n");
+
+	psy_cfg.drv_data = chip;
+	psy_cfg.of_node = pdev->dev.of_node;
+
+	/* Power supply */
+	chip->batt_psy =
+		devm_power_supply_register(chip->dev, &batt_psy_desc, &psy_cfg);
+	if (IS_ERR(chip->batt_psy))
+		return dev_err_probe(chip->dev, PTR_ERR(chip->batt_psy),
+				     "Failed to register power supply\n");
+
+	/* Battery info */
+	ret = power_supply_get_battery_info(chip->batt_psy, &chip->batt_info);
+	if (ret)
+		return dev_err_probe(chip->dev, ret,
+				     "Failed to get battery info\n");
+
+	platform_set_drvdata(pdev, chip);
+
+	return 0;
+}
+
+static const struct of_device_id qcom_qg_of_match[] = {
+	{ .compatible = "qcom,pm6150-qg", },
+	{ /* sentinel */ },
+};
+MODULE_DEVICE_TABLE(of, qcom_qg_of_match);
+
+static struct platform_driver qcom_qg_driver = {
+	.driver = {
+		.name = "qcom,qcom_qg",
+		.of_match_table = qcom_qg_of_match,
+	},
+	.probe = qcom_qg_probe,
+};
+
+module_platform_driver(qcom_qg_driver);
+
+MODULE_AUTHOR("Danila Tikhonov <danila@jiaxyga.com>");
+MODULE_DESCRIPTION("Qualcomm PMIC QGauge (QG) driver");
+MODULE_LICENSE("GPL");
